### PR TITLE
Bug 1690456: validate directory provided by --metadata-url during bootstrap

### PR DESCRIPTION
**Pull request by hmlanigan**
If the directory ends in 'tools' or images', assume
that only the respective value will be used.  If a high level
directory is listed, check for tools and images in it.  If no
tools are listed, use tools from streams.canonical.com.

## Description of change

Attempting to anticipate what users intend with bootstrap --metadata-source.
* If the directory doesn't exist, fail quickly so the user can retry.
* If the directory does exist, look for tools and images directories inside.
  * If tools does exist, replace the default tool location.
  * If tools doesn't exist, leave the default tools as simple streams.canonical.com....
* If the directory ends in 'images' assume that only image metadata will be found
* If the directory ends in 'tools' assume that only tools metadata will be found

The contents of tools and images are validated further on.

## QA steps

Bootstrap openstack with arm64 nova compute nodes from a host machine with an a
different processor (arm64), don't use a nightly build to test as tools for nightly are not
on streams.canonical.com. - to verify bug fix.

Unit tests will be updated to verify the listed possibilities.

## Documentation changes

N/A

## Bug reference

https://bugs.launchpad.net/juju/+bug/1690456

### CODE REVIEW

File: environs/bootstrap/bootstrap.go [hunk 1]
```diff
@@ -659,15 +659,33 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
 // for tools syncing, and adds an image metadata source after verifying
 // the contents.
 func setPrivateMetadataSources(metadataDir string) ([]*imagemetadata.ImageMetadata, error) {
-	logger.Infof("Setting default tools and image metadata sources: %s", metadataDir)
-	tools.DefaultBaseURL = metadataDir
+	if _, err := os.Stat(metadataDir); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, errors.Annotate(err, "cannot access metadata directory")
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: "... access simplestreams metadata ..."

File: environs/bootstrap/bootstrap.go [hunk 2]
```diff
@@ -659,15 +659,33 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+		}
+		logger.Infof("metadata source: %s, not found", metadataDir)
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: We should return a not found error here so it bubbles up and aborts bootstrap
return errors.NotFoundf("simplestreams metadata source directory %v", metadataDir)

File: environs/bootstrap/bootstrap.go [hunk 3]
```diff
@@ -659,15 +659,33 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+		return nil, nil
+	}
+
+	toolsMetadataDir := filepath.Join(metadataDir, storage.BaseToolsPath)
+	if _, err := os.Stat(toolsMetadataDir); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, errors.Annotate(err, "cannot access tools metadata")
+		}
+		logger.Infof("default tools metadata source: %s", tools.DefaultBaseURL)
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: no need to log at info, just debug

File: environs/bootstrap/bootstrap.go [hunk 4]
```diff
@@ -659,15 +659,33 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+	} else {
+		logger.Infof("setting default tools metadata source: %s", toolsMetadataDir)
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: ditto about debug

File: environs/bootstrap/bootstrap.go [hunk 5]
```diff
@@ -659,15 +659,33 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+		tools.DefaultBaseURL = metadataDir
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: we only want to do this if metadataDir ends with /tools
so the logic for this block should be:

toolsMetadataDir = metadataDir
if not toolsMetadataDir endswith /tools then
  toolsMetadataDir += /tools
if toolsMetadataDir exists then
  tools.DefaultBaseURL = toolsMetadataDir

File: environs/bootstrap/bootstrap.go [hunk 6]
```diff
@@ -659,15 +659,33 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+	}
 
 	imageMetadataDir := filepath.Join(metadataDir, storage.BaseImagesPath)
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: we want the same logic as for tools above.
ie
the user can specify --metadata-source as one of:

somedir
somedir/tools
somedir/images

and the logic should Do The Right Thing

File: environs/bootstrap/bootstrap.go (hunk 1 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 2 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 3 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 4 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 5 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 6 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

**Comment by hmlanigan** [comment]
working on tests now.

File: cmd/juju/commands/bootstrap_test.go [hunk 7]
```diff
@@ -981,11 +981,11 @@ func (s *BootstrapSuite) TestInvalidLocalSource(c *gc.C) {
...
 	stderr := cmdtesting.Stderr(ctx)
 	c.Check(stderr, gc.Matches,
-		"(.|\n)*Looking for packaged Juju agent version 1\\.2\\.0 for amd64\n"+
-			"No packaged binary found, preparing local Juju agent binary(.|\n)*",
+		"Creating Juju controller \"devcontroller\" on dummy/dummy\n"+
+			"Looking for packaged Juju agent version 1.2.0 for amd64\n",
 	)
 	c.Check(s.tw.Log(), jc.LogMatches, []jc.SimpleMessage{
-		{loggo.ERROR, "failed to bootstrap model: cannot package bootstrap agent binary: no agent binaries for you"},
+		{loggo.ERROR, "failed to bootstrap model: no matching tools available"},
```
Comment (jameinel) [inline cmd/juju/commands/bootstrap_test.go]: 'agent binaries' is the preferred name to 'matching tools'.
Mark especially doesn't like the word 'tools' and we should avoid using it in messages to users.

File: environs/bootstrap/bootstrap.go [hunk 8]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
 	return v1.Compare(v2) == 0
 }
 
-// setPrivateMetadataSources sets the default tools metadata source
-// for tools syncing, and adds an image metadata source after verifying
-// the contents.
+// setPrivateMetadataSources verifys the metadataDir and uses it to
+// set the default tools metadata source for tools syncing, and adds
```
Comment (jameinel) [inline environs/bootstrap/bootstrap.go]: s/tools/agent binaries/

I realize we're currently looking for a directory ending in 'tools' but telling people 'we find agent binaries in a directory named tools' is still clearer what we are actually looking for. 'tools' could be anything (such as helper commands, eg juju-metadata, etc).

File: environs/bootstrap/bootstrap.go [hunk 9]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+	}
+	if _, err := os.Stat(toolsMetadataDir); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, errors.Annotate(err, "cannot access tools metadata")
+		}
+		logger.Debugf("default tools metadata source: %s", tools.DefaultBaseURL)
+	} else {
+		if ending == storage.BaseToolsPath {
+			tools.DefaultBaseURL = filepath.Dir(metadataDir)
+			logger.Debugf("setting default tools metadata source: %s", tools.DefaultBaseURL)
```
Comment (jameinel) [inline environs/bootstrap/bootstrap.go]: agent binaries

especially given the config option would be "agent-metadata-url"

**Comment by wallyworld** [comment]
@jameinel the reason for one directory is that the user can choose to put there whatever necessary metadata is needed. They may want to provide image metadata, agent metadata, or both. They do this by using sub directories "images" or "tools" <- yes, "tools" is outdated. We should add support for "agents" as well. Regardless, the --metadata-source dir is meant to be the top level dir which contains sub dirs.

The issue addressed here is partly a bug (Juju wasn't checking the validity of the dir supplied), but also a usability issue - people have been specifying a full path to "sourcedir/images" instead of "sourcedir". So part of the change here is to support DWIM.

File: cmd/juju/commands/bootstrap_test.go (hunk 7 above)
Comment (wallyworld) [inline cmd/juju/commands/bootstrap_test.go]: There's still legacy error messages floating around - agree it would be good to fix those as a driveby. We should also support looking for "agents" as well as "tools" as a sub dir under metadatasource dir.

Comment (hmlanigan) [inline cmd/juju/commands/bootstrap_test.go]: The wording update in this specific case will be addressed in a follow-on PR, as will
looking for an 'agents' directory.

File: environs/bootstrap/bootstrap.go (hunk 8 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: updated the log messages to use agent binary instead of tools

**Comment by hmlanigan** [comment]
!!testme!!

File: environs/bootstrap/bootstrap.go [hunk 10]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
 }
 
-// setPrivateMetadataSources sets the default tools metadata source
-// for tools syncing, and adds an image metadata source after verifying
-// the contents.
+// setPrivateMetadataSources verifys the metadataDir and uses it to
+// set the default agent binaries metadata source for agent binary
+// syncing, and adds an image metadata source after verifying the
+// contents. If the directory ends in tools, only the default tools
+// metadata source will be set.  Same for images.
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: remove extra space

File: environs/bootstrap/bootstrap.go [hunk 11]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
 	return v1.Compare(v2) == 0
 }
 
-// setPrivateMetadataSources sets the default tools metadata source
-// for tools syncing, and adds an image metadata source after verifying
-// the contents.
+// setPrivateMetadataSources verifys the metadataDir and uses it to
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: verifies the specified metadataDir exists

File: environs/bootstrap/bootstrap.go [hunk 12]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+// set the default agent binaries metadata source for agent binary
+// syncing, and adds an image metadata source after verifying the
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: s/syncing//

File: environs/bootstrap/bootstrap.go [hunk 13]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+	agentBinaryMetadataDir := metadataDir
+	ending := filepath.Base(agentBinaryMetadataDir)
+	if ending != storage.BaseToolsPath {
+		agentBinaryMetadataDir = filepath.Join(metadataDir, storage.BaseToolsPath)
+	}
+	if _, err := os.Stat(agentBinaryMetadataDir); err != nil {
+		if !os.IsNotExist(err) {
+			return nil, errors.Annotate(err, "cannot access agent binary metadata")
+		}
+		logger.Debugf("default agent binary metadata source: %s", tools.DefaultBaseURL)
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: can we reword to "no agent directory found, using default agent binary....."

File: environs/bootstrap/bootstrap.go [hunk 14]
```diff
@@ -655,19 +655,52 @@ func isCompatibleVersion(v1, v2 version.Number) bool {
...
+	} else {
+		if ending == storage.BaseToolsPath {
+			tools.DefaultBaseURL = filepath.Dir(metadataDir)
+			logger.Debugf("setting default agent binary metadata source: %s", tools.DefaultBaseURL)
+			return nil, nil
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap.go]: can we add a comment explaining why it's ok to exit early here

File: environs/bootstrap/bootstrap_test.go [hunk 15]
```diff
@@ -931,6 +934,106 @@ func (s *bootstrapSuite) TestBootstrapMetadata(c *gc.C) {
...
+	})
+	c.Assert(err, gc.NotNil)
+	c.Assert(err, gc.ErrorMatches, fmt.Sprintf("simplestreams metadata source: %s not found", nonExistentFileName))
+}
+
+// TestBootstrapMetadataImagesNoTools tests 2 cases:
+// juju bootstrap --metadata-source <dir>
+// juju bootstrap --metadata-source <dir>/images
+// where <dir>/tools doesn't exist
+func (s *bootstrapSuite) TestBootstrapMetadataImagesNoTools(c *gc.C) {
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap_test.go]: this should either be 2 tests, or a single test inside a for loop
metadataDir, _ := createImageMetadata(c)
for _, suffix := range []string{"", "images"} {
  sourceDir := filepath.Join(metadataDir, suffix)
  do test stuff...
}

File: environs/bootstrap/bootstrap_test.go [hunk 16]
```diff
@@ -931,6 +934,106 @@ func (s *bootstrapSuite) TestBootstrapMetadata(c *gc.C) {
...
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(datasources, gc.HasLen, 3)
+	c.Assert(datasources[0].Description(), gc.Equals, "bootstrap metadata")
+}
+
+// TestBootstrapMetadataToolsNoImages tests 2 cases:
+// juju bootstrap --metadata-source <dir>
+// juju bootstrap --metadata-source <dir>/tools
+// where <dir>/images doesn't exist
+func (s *bootstrapSuite) TestBootstrapMetadataToolsNoImages(c *gc.C) {
```
Comment (wallyworld) [inline environs/bootstrap/bootstrap_test.go]: ditto

File: environs/bootstrap/bootstrap.go (hunk 11 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 12 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 10 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 13 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap.go (hunk 14 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap.go]: done

File: environs/bootstrap/bootstrap_test.go (hunk 15 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap_test.go]: done

File: environs/bootstrap/bootstrap_test.go (hunk 16 above)
Comment (hmlanigan) [inline environs/bootstrap/bootstrap_test.go]: done

**Comment by hmlanigan** [comment]
manual QA has been done on the last changes.

**Comment by hmlanigan** [comment]
$$merge$$

**Comment by hmlanigan** [comment]
$$merge$$
